### ISSUE TITLE
smartics generator includes redundant module definitions

### ISSUE THREAD

**Issue by tdiesler** [wildfly-extras/wildfly-camel#184]
Currently there is no mechanism that guarantees the correct wiring to existing wildfly modules. As a result, redundant/incorrect module definitions are being generated.

org/apache/logging for example

```
[tdiesler@macbook modules]$ tar tzf target/wildfly-camel-modules-2.1.0-SNAPSHOT.tar.gz | grep jar | sort
modules/ca/uhn/hapi/base/main/hapi-base-2.2.jar
modules/com/jcraft/jsch/main/jsch-0.1.50.jar
modules/io/github/rometools/rome/main/rome-1.0.jar
modules/io/netty/netty-buffer/main/netty-buffer-4.0.23.Final.jar
modules/io/netty/netty-codec/main/netty-codec-4.0.23.Final.jar
modules/io/netty/netty-common/main/netty-common-4.0.23.Final.jar
modules/io/netty/netty-handler/main/netty-handler-4.0.23.Final.jar
modules/io/netty/netty-transport/main/netty-transport-4.0.23.Final.jar
modules/net/sf/saxon/he/main/Saxon-HE-9.5.1-5.jar
modules/org/apache/abdera/core/main/abdera-core-1.1.3.jar
modules/org/apache/abdera/i18n/main/abdera-i18n-1.1.3.jar
modules/org/apache/abdera/parser/main/abdera-parser-1.1.3.jar
modules/org/apache/activemq/camel/main/activemq-camel-5.10.0.jar
modules/org/apache/activemq/main/activemq-broker-5.10.0.jar
modules/org/apache/activemq/main/activemq-client-5.10.0.jar
modules/org/apache/activemq/main/activemq-jms-pool-5.10.0.jar
modules/org/apache/activemq/main/activemq-kahadb-store-5.10.0.jar
modules/org/apache/activemq/main/activemq-openwire-legacy-5.10.0.jar
modules/org/apache/activemq/main/activemq-pool-5.10.0.jar
modules/org/apache/activemq/main/activemq-protobuf-1.1.jar
modules/org/apache/activemq/main/activemq-ra-5.10.0.jar
modules/org/apache/activemq/main/activemq-spring-5.10.0.jar
modules/org/apache/camel/component/atom/main/camel-atom-2.14.0.jar
modules/org/apache/camel/component/bindy/main/camel-bindy-2.14.0.jar
modules/org/apache/camel/component/cdi/main/camel-cdi-2.14.0.jar
modules/org/apache/camel/component/cxf/main/camel-cxf-2.14.0.jar
modules/org/apache/camel/component/cxf/main/camel-cxf-transport-2.14.0.jar
modules/org/apache/camel/component/ftp/main/camel-ftp-2.14.0.jar
modules/org/apache/camel/component/hl7/main/camel-hl7-2.14.0.jar
modules/org/apache/camel/component/jaxb/main/camel-jaxb-2.14.0.jar
modules/org/apache/camel/component/jms/main/camel-jms-2.14.0.jar
modules/org/apache/camel/component/jmx/main/camel-jmx-2.14.0.jar
modules/org/apache/camel/component/jpa/main/camel-jpa-2.14.0.jar
modules/org/apache/camel/component/lucene/main/camel-lucene-2.14.0.jar
modules/org/apache/camel/component/mail/main/camel-mail-2.14.0.jar
modules/org/apache/camel/component/mina2/main/camel-mina2-2.14.0.jar
modules/org/apache/camel/component/mqtt/main/camel-mqtt-2.14.0.jar
modules/org/apache/camel/component/mvel/main/camel-mvel-2.14.0.jar
modules/org/apache/camel/component/netty4/main/camel-netty4-2.14.0.jar
modules/org/apache/camel/component/ognl/main/camel-ognl-2.14.0.jar
modules/org/apache/camel/component/quartz/main/camel-quartz-2.14.0.jar
modules/org/apache/camel/component/rss/main/camel-rss-2.14.0.jar
modules/org/apache/camel/component/saxon/main/camel-saxon-2.14.0.jar
modules/org/apache/camel/component/script/main/camel-script-2.14.0.jar
modules/org/apache/camel/component/soap/main/camel-soap-2.14.0.jar
modules/org/apache/camel/component/sql/main/camel-sql-2.14.0.jar
modules/org/apache/camel/component/weather/main/camel-weather-2.14.0.jar
modules/org/apache/camel/core/main/camel-core-2.14.0.jar
modules/org/apache/camel/core/xml/main/camel-core-xml-2.14.0.jar
modules/org/apache/camel/spring/main/camel-spring-2.14.0.jar
modules/org/apache/commons/codec/main/commons-codec-1.9.jar
modules/org/apache/commons/logging/main/commons-logging-1.0.3.jar
modules/org/apache/commons/net/main/commons-net-3.3.jar
modules/org/apache/commons/pool/main/commons-pool-1.6.jar
modules/org/apache/deltaspike/cdictrl/api/main/deltaspike-cdictrl-api-1.0.2.jar
modules/org/apache/deltaspike/core/api/main/deltaspike-core-api-1.0.2.jar
modules/org/apache/deltaspike/core/impl/main/deltaspike-core-impl-1.0.2.jar
modules/org/apache/geronimo/specs/geronimo-javamail_1/4_spec/main/geronimo-javamail_1.4_spec-1.7.1.jar
modules/org/apache/geronimo/specs/geronimo-osgi-locator/main/geronimo-osgi-locator-1.0.jar
modules/org/apache/geronimo/specs/geronimo-osgi-registry/main/geronimo-osgi-registry-1.0.jar
modules/org/apache/jakarta/regexp/main/jakarta-regexp-1.4.jar
modules/org/apache/james/apache-mime4j-core/main/apache-mime4j-core-0.7.2.jar
modules/org/apache/lucene/analyzers-common/main/lucene-analyzers-common-4.6.1.jar
modules/org/apache/lucene/core/main/lucene-core-4.6.1.jar
modules/org/apache/lucene/queries/main/lucene-queries-4.6.1.jar
modules/org/apache/lucene/queryparser/main/lucene-queryparser-4.6.1.jar
modules/org/apache/lucene/sandbox/main/lucene-sandbox-4.6.1.jar
modules/org/apache/mina/core/main/mina-core-2.0.7.jar
modules/org/apache/servicemix/bundles/javassist/main/org.apache.servicemix.bundles.javassist-3.12.1.GA_3.jar
modules/org/apache/servicemix/bundles/ognl/main/org.apache.servicemix.bundles.ognl-3.0.8_1.jar
modules/org/apache/ws/commons/axiom/axiom-api/main/axiom-api-1.2.14.jar
modules/org/apache/ws/commons/axiom/axiom-impl/main/axiom-impl-1.2.14.jar
modules/org/apache/xbean/spring/main/xbean-spring-3.16.jar
modules/org/codehaus/jackson/jackson-core-asl/main/jackson-core-asl-1.9.12.jar
modules/org/codehaus/jackson/jackson-mapper-asl/main/jackson-mapper-asl-1.9.12.jar
modules/org/codehaus/woodstox/stax2-api/main/stax2-api-3.1.1.jar
modules/org/codehaus/woodstox/woodstox-core-asl/main/woodstox-core-asl-4.1.4.jar
modules/org/fusesource/hawtbuf/main/hawtbuf-1.10.jar
modules/org/fusesource/hawtdispatch/main/hawtdispatch-1.21.jar
modules/org/fusesource/hawtdispatch/transport/main/hawtdispatch-transport-1.21.jar
modules/org/fusesource/mqtt-client/main/mqtt-client-1.10.jar
modules/org/jaxen/main/jaxen-1.1.1.jar
modules/org/jboss/gravia/main/gravia-provision-1.1.0.Beta44.jar
modules/org/jboss/gravia/main/gravia-repository-1.1.0.Beta44.jar
modules/org/jboss/gravia/main/gravia-resolver-1.1.0.Beta44.jar
modules/org/jboss/gravia/main/gravia-resource-1.1.0.Beta44.jar
modules/org/jboss/gravia/main/gravia-runtime-api-1.1.0.Beta44.jar
modules/org/jboss/gravia/main/gravia-runtime-embedded-1.1.0.Beta44.jar
modules/org/jboss/gravia/main/org.apache.felix.configadmin-1.8.0.jar
modules/org/jboss/gravia/main/org.apache.felix.http.bridge-2.2.1.jar
modules/org/jboss/gravia/main/org.apache.felix.log-1.0.1.jar
modules/org/jboss/gravia/main/org.apache.felix.metatype-1.0.8.jar
modules/org/jboss/gravia/main/org.apache.felix.scr-1.6.2.jar
modules/org/mvel/mvel2/main/mvel2-2.2.1.Final.jar
modules/org/osgi/core/main/org.osgi.core-4.3.1.jar
modules/org/osgi/enterprise/main/org.osgi.enterprise-5.0.0.jar
modules/org/quartz-scheduler/quartz/main/quartz-1.8.6.jar
modules/org/springframework/aop/main/aopalliance-1.0.jar
modules/org/springframework/aop/main/spring-aop-3.2.11.RELEASE.jar
modules/org/springframework/beans/main/spring-beans-3.2.11.RELEASE.jar
modules/org/springframework/context/main/spring-context-3.2.11.RELEASE.jar
modules/org/springframework/core/main/spring-core-3.2.8.RELEASE.jar
modules/org/springframework/expression/main/spring-expression-3.2.11.RELEASE.jar
modules/org/springframework/jdbc/main/spring-jdbc-3.2.11.RELEASE.jar
modules/org/springframework/jms/main/spring-jms-3.2.11.RELEASE.jar
modules/org/springframework/orm/main/spring-orm-3.2.11.RELEASE.jar
modules/org/springframework/transaction/main/spring-tx-3.2.11.RELEASE.jar
```

We support every module we ship. Therefore, its important that we only ship what we can support ;-)

* Duplicate module definitions (possibly with multiple versions) must be avoided
* License constraints must be applied

If this cannot be automated, we may need to use smartics to generate the next module definition "suggestion" and have an additional review + copy step.

An alternative may be to maintain an list like above in scm and make sure that every change to the list must manually be endorsed.

**Comment by jamesnetherton** [comment]
I've done a bit more work on this and the results can be found at commit a1d124eb0ab2ed1fd716035a1fa0e5ce32c25677. There's also some additional documentation [here](https://github.com/jamesnetherton/wildfly-camel/tree/184-better-dependency-checks/modules/scripts).

I went through several iterations of trying to get something working with standalone scripts and attempting to modify the smartics plugin.

I've settled on a simple Groovy script which does some comparisons between the wildfly camel module dependencies and those supplied with the wildfly app server. Using Groovy at least keeps the build cross platform.

It's not 100% foolproof. Since it's comparing Maven group / artifact ids, there are scenarios where duplicate dependencies can slip though. E.g with netty where WF uses the 'super' netty-all jar but WF camel uses a selection of individual netty jars. In this instance it'll be for code review to rectify.
